bundles/io/org.openhab.io.gcal/src/main/java/org/openhab/io/gcal/internal/GCalEventDownloader.java

```diff
@@ -533,4 +605,136 @@ public void updated(Dictionary<String, ?> config) throws ConfigurationException
+                mapData.put("client_secret", client_secret);
+                mapData.put("code", device.device_code);
+                mapData.put("grant_type", "http://oauth.net/grant_type/device/1.0");
+
+                content = new UrlEncodedContent(mapData);
+                postRequest = requestFactory
+                        .buildPostRequest(new GenericUrl("https://accounts.google.com/o/oauth2/token"), content);
+
+                DeviceToken deviceToken;
+                do {
+                    deviceToken = postRequest.execute().parseAs(DeviceToken.class);
+
+                    if (deviceToken.access_token != null) {
+                        if (logger.isDebugEnabled()) {
+                            logger.debug("Got access token");
+                            logger.debug("device access token: " + deviceToken.access_token);
+                            logger.debug("device token_type: " + deviceToken.token_type);
+                            logger.debug("device refresh_token: " + deviceToken.refresh_token);
+                            logger.debug("device expires_in: " + deviceToken.expires_in);
+                        }
+                        break;
+                    }
+                    logger.debug("waiting for {} seconds", device.interval);
+                    Thread.sleep(device.interval * 1000);
+
+                } while (true);
+
+                StoredCredential dataCredential = new StoredCredential();
+                dataCredential.setAccessToken(deviceToken.access_token);
+                dataCredential.setRefreshToken(deviceToken.refresh_token);
+                dataCredential.setExpirationTimeMilliseconds((long) deviceToken.expires_in * 1000);
+
+                datastore.set(TOKEN_STORE_USER_ID, dataCredential);
+
+                credential = loadCredential(TOKEN_STORE_USER_ID, datastore);
+            }
+        } catch (Exception e) {
+            logger.warn("getCredential got exception: " + e.getMessage());
+        }
+
+        return credential;
+    }
+
+    private static Credential loadCredential(String userId, DataStore<StoredCredential> credentialDataStore)
+            throws IOException {
+        Credential credential = newCredential(userId, credentialDataStore);
+        if (credentialDataStore != null) {
+            StoredCredential stored = credentialDataStore.get(userId);
+            if (stored == null) {
+                return null;
+            }
+            credential.setAccessToken(stored.getAccessToken());
+            credential.setRefreshToken(stored.getRefreshToken());
+            credential.setExpirationTimeMilliseconds(stored.getExpirationTimeMilliseconds());
+
+            logger.debug("Loaded credential");
+            logger.debug("device access token: " + stored.getAccessToken());
+            logger.debug("device refresh_token: " + stored.getRefreshToken());
+            logger.debug("device expires_in: " + stored.getExpirationTimeMilliseconds());
```

Comment:
Use parameterized logging, like
```
logger.debug("device expires_in: {}", stored.getExpirationTimeMilliseconds())
```
for all logger calls.  This prevents unnecessary object creation and garbage collection.